BUG: spatial.HalfspaceIntersection: fix `add_halfspaces` batch size support

rsbowen opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
This [stemmed](https://github.com/scipy/scipy/pull/20035#discussion_r1951830860) from the investigation of, but does not close, https://github.com/scipy/scipy/issues/19865. Please let me know if I should also file a separate bug for this issue.

#### What does this implement/fix?
`add_halfspaces` fails with an array-size issue unless the number of halfspaces being added is `1`, or (I think) `ndim`.  The only places this function was tested is [here](https://github.com/scipy/scipy/blob/main/scipy/spatial/tests/test_qhull.py#L1135), where only a single halfspace is added at a time. I added a simple test and fixed the broadcasting.

#### Additional information
This is my first PR to scipy or indeed to any open source project with this much complexity. Please LMK if I have missed anything.

rsbowen committed:
Add a (failing) test that checks we can add halfspaces with various batch sizes

rsbowen committed:
Fix dims for dists in the case that there is more than one halfspace being added

rsbowen committed:
Add correctness, not just doesnt-error, test

tylerjereddy commented on scipy/spatial/tests/test_qhull.py:
```diff
@@ -1186,6 +1186,38 @@ def test_cube(self):
+        close = np.all(close, axis=-1)
+
+        assert (np.count_nonzero(close, axis=0) == 1).all()
+        assert (np.count_nonzero(close, axis=1) == 1).all()
```
I wonder if it might be slightly clearer to sort the actual and expected arrays lexically and then compare them directly  with `assert_array_equal`:

```diff
--- a/scipy/spatial/tests/test_qhull.py
+++ b/scipy/spatial/tests/test_qhull.py
@@ -1212,12 +1212,9 @@ class Test_HalfspaceIntersection:
         actual_intersections = hs.intersections
         # They may be in any order, so just check that under some permutation 
         # expected=actual.
-        close = np.isclose(actual_intersections[np.newaxis],
-                           expected_intersections[:, np.newaxis])
-        close = np.all(close, axis=-1)
-
-        assert (np.count_nonzero(close, axis=0) == 1).all()
-        assert (np.count_nonzero(close, axis=1) == 1).all()
+        ind1 = np.lexsort((actual_intersections[:, 1], actual_intersections[:, 0]))
+        ind2 = np.lexsort((expected_intersections[:, 1], expected_intersections[:, 0]))
+        assert_array_equal(actual_intersections[ind1], expected_intersections[ind2])
```

rsbowen replied:
Good idea, [8e7360a](https://github.com/scipy/scipy/pull/22693/commits/8e7360a6bc450978d35ef64b703d6a0a48c6beb1)

rsbowen committed:
Make the feasible point not the origin so that the dist array is not all the same

tylerjereddy reviewed (commented):
This seems sensible and will likely unblock the other bug fix PR indeed.

It makes sense that the intersection of the two convex sets (squares) is the internal one:

<details>

<summary> Sample Plot of the two convex sets (squares) </summary>

![image](https://github.com/user-attachments/assets/714c6846-db5c-4a06-8ea7-8e5ff7638816)

</details>

One other thing that might be useful to do is update the docstring for `add_halfspaces`  so that the `halfspaces` parameter has a shape that is a bit more specific, like in the parent class:

![image](https://github.com/user-attachments/assets/47a51ff8-b306-4486-a4cd-85718414b38c)

vs. 

![image](https://github.com/user-attachments/assets/f10dac5c-6cb3-4c2f-8c3b-c88da12e86c9)

I double checked that the test fails before/passes after the patch so that's good.

The only other minor comment I'd make is that we try to format our commit messages per the following, if you continue to contribute: https://docs.scipy.org/doc/scipy/dev/contributor/development_workflow.html#writing-the-commit-message

rsbowen commented:
There is a little more to it. In existing code [here](https://github.com/scipy/scipy/blob/main/scipy/spatial/_qhull.pyx#L459) the size of `dists` is `(num_new_halfspaces,)` and of `points[:,:-1]` is `(num_new_halfspaces, ndim)`. That's why it works when `ndim` is 1. The broadcasting is also legal when `ndim = num_new_halfspaces`, but this does the division along the wrong axis. So you can get incorrect results

```python
big_square = np.array([[ 1.,  0., -2.],
                        [-1.,  0., -2.],
                        [ 0.,  1., -2.],
                        [ 0., -1., -2.]])

small_square = np.array([[ 1.,  0., -1.],
                          [-1.,  0., -1.],
                          [ 0.,  1., -1.],
                          [ 0., -1., -1.]])

hs = HalfspaceIntersection(big_square, np.array([0.3141, 0.2718]), incremental=True)
hs.add_halfspaces(small_square[:2])
hs.add_halfspaces(small_square[2:])
hs.intersections

array([[ 1.    ,  1.5436],
       [-0.3718,  1.5436],
       [ 1.    , -1.    ],
       [-0.3718, -1.    ]])
```

This happens in the (probably common) case that `dists` has more than one value. However the test I wrote doesn't check this, because it chooses the origin as the initial feasible point, which has the same signed distance to every point. So, I updated the test to use a different feasible point, `[0.3141, 0.2718]` as a center; that way, this test also acts a regression test for that bug ([51a9ae9](https://github.com/scipy/scipy/pull/22693/commits/51a9ae941e79025055967e5e943f4390ca9452d9))

rsbowen committed:
TST: spatial.HalfspaceIntersection: use sorting and assert_allclose

This is easier to read than the previous permutation-matrix solution

rsbowen committed:
DOC: spatial.HalfspaceIntersection: make shape and format more explicit

Update the docs of add_halfspaces to look more like the constructor, to
make it crystal clear what the shapes should be, as well as the fact
that these new half spaces need to contain the original feasible point.

rsbowen commented:
> One other thing that might be useful to do is update the docstring for `add_halfspaces` so that the `halfspaces` parameter has a shape that is a bit more specific, like in the parent class:

[51e3e7f](https://github.com/scipy/scipy/pull/22693/commits/51e3e7f8a5c61f26c2ced20dd2bf6b64b5981a0a)

tylerjereddy commented:
sounds good, let's let the CI run then probably merge soon

rsbowen commented:
Not sure what to do about the CI failure. Could it be CI flakyness? ` python dev.py test -t scipy/sparse` passes on my machine with `13274 passed, 1996 skipped, 4808 deselected, 57 xfailed, 3 xpassed in 87.74s (0:01:27)`

tylerjereddy committed:
DOC: PR 22693 revisions

* `add_halfspaces()` docstring cleanup to avoid "Input halfspaces."
showing up as a parameter.

[ci skip]

tylerjereddy commented:
> Not sure what to do about the CI failure. 

It is not related. For example, it also shows up in a recent PR that only changes a docstring: https://github.com/scipy/scipy/pull/22698

tylerjereddy commented:
I pushed in a small commit to fix up a docstring blemish where a few words were showing up as an "extra parameter:"

![image](https://github.com/user-attachments/assets/5f10ea65-1b90-4542-86a1-f8614fe10d48)

to 

![image](https://github.com/user-attachments/assets/0ab5dc2c-e7f6-467d-b9e9-30758535c07e)

tylerjereddy commented:
@rsbowen thanks for your first contribution to SciPy. I'll ping you for a review on the PR that started all this after I rebase/update it.
